test/api/model/web_auth_test.dart

```diff
@@ -0,0 +1,97 @@
+import 'package:checks/checks.dart';
+import 'package:convert/convert.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:zulip/api/model/web_auth.dart';
+
+import '../../example_data.dart' as eg;
+
+void main() {
+  group('WebAuthPayload', () {
+    const otp = '186f6d085a5621ebaf1ccfc05033e8acba57dae03f061705ac1e58c402c30a31';
+    final encryptedApiKey = debugEncodeApiKey(eg.selfAccount.apiKey, otp);
+    final wellFormed = Uri.parse(
+      'zulip://login?otp_encrypted_api_key=$encryptedApiKey'
+        '&email=self%40example&user_id=1&realm=https%3A%2F%2Fchat.example%2F');
+
+    test('basic happy case', () {
+      final payload = WebAuthPayload.parse(wellFormed);
+      check(payload)
+        ..otpEncryptedApiKey.equals(encryptedApiKey)
+        ..email.equals('self@example')
+        ..userId.equals(1)
+        ..realm.equals(Uri.parse('https://chat.example/'));
+      check(payload.decodeApiKey(otp)).equals(eg.selfAccount.apiKey);
+    });
+
+    // TODO(server-5) remove
+    test('legacy: no userId', () {
+      final queryParams = {...wellFormed.queryParameters}..remove('user_id');
+      final payload = WebAuthPayload.parse(
+        wellFormed.replace(queryParameters: queryParams));
+      check(payload)
+        ..otpEncryptedApiKey.equals(encryptedApiKey)
+        ..email.equals('self@example')
+        ..userId.isNull()
+        ..realm.equals(Uri.parse('https://chat.example/'));
+      check(payload.decodeApiKey(otp)).equals(eg.selfAccount.apiKey);
+    });
+
+    test('parse fails when an expected field is missing', () {
+      final queryParams = {...wellFormed.queryParameters}..remove('email');
+      final input = wellFormed.replace(queryParameters: queryParams);
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when otp_encrypted_api_key is wrong length', () {
+      final queryParams = {...wellFormed.queryParameters}
+        ..['otp_encrypted_api_key'] = 'asdf';
+      final input = wellFormed.replace(queryParameters: queryParams);
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when host is not "login"', () {
+      final input = wellFormed.replace(host: 'foo');
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when scheme is not "zulip"', () {
+      final input = wellFormed.replace(scheme: 'https');
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('decodeApiKey fails when otp is wrong length', () {
+      final payload = WebAuthPayload.parse(wellFormed);
+      check(() => payload.decodeApiKey('asdf')).throws<FormatException>();
+    });
+  });
+
+  group('generateOtp', () {
+    test('smoke, and check all 256 byte values are used', () {
+      const n = 216;
```

Comment:
Let's make the principle here a bit more explicit — helpful for anyone thinking of adding another probabilistic test and finding this as an example:
```suggestion
    test('smoke, and check all 256 byte values are used', () {
      // This is a probabilistic test.  We've chosen `n` so that when the test
      // should pass, the probability it fails is < 1e-9.  See analysis below.
      const n = 216;
```